### PR TITLE
add plumbing for aws secret manager

### PR DESCRIPTION
**Pull request by alexgallotta**
PR contains:
- a few refactors and moving around of unused files (bottlecap local runner)
- creating script for quick build & publish of only bottlecap, reusing published agent. This would ensure that the same agent version is used for different bottlecaps. Suffix is hardcoded for now
- secret manager first implementation, blocking. It will panic if API key ARN is set but can't be decrypted 
```
2024-05-16T19:32:55.219Z	[DEBUG] invoke{service=secretsmanager operation=GetSecretValue}: apply_configuration: timeout settings for this operation: TimeoutConfig { co
...
2024-05-16T19:32:55.277Z	[DEBUG] invoke{service=secretsmanager operation=GetSecretValue}: try_op: try_attempt: entering 'after deserialization' phase
2024-05-16T19:32:55.277Z	[DEBUG] invoke{service=secretsmanager operation=GetSecretValue}: try_op: attempt #1 succeeded, no retry necessary
2024-05-16T19:32:55.277Z	[DEBUG] invoke{service=secretsmanager operation=GetSecretValue}: try_op: a retry is either unnecessary or not possible, exiting attempt loop
```

to decrypt at the first try it seems to take around 277-219 = 58ms. Average over time is around ~25ms, that might benefit for warm/caching of the key somewhere

### CODE REVIEW

File: bottlecap-run/runBottlecap.sh
Comment (duncanista) [inline bottlecap-run/runBottlecap.sh:1]: Could we move this changes to another PR if they are not relevant to Secrets Manager?

File: bottlecap/Cargo.toml [hunk 1]
```diff
@@ -16,13 +18,13 @@ protobuf = { version = "3.4.0", default-features = false }
...
 thiserror = { version = "1.0.58", default-features = false }
-tracing = {version = "0.1.40", default-features = false }
+tracing = { version = "0.1.40", default-features = false }
 tracing-core = { version = "0.1.32", default-features = false }
 tracing-log = { version = "0.2.0", default-features = false, features = ["std", "log-tracer"] }
 tracing-subscriber = { version = "0.3.18", default-features = false, features = ["fmt", "env-filter"] }
 ureq = { version = "2.9.7", features = ["tls", "json"], default-features = false }
 ustr = { version = "1.0.0", default-features = false }
-
+tokio = { version = "1.37.0" }
```
Comment (duncanista) [inline bottlecap/Cargo.toml:27]: Are we sure we want everything included here?

File: bottlecap/src/bin/bottlecap/main.rs [hunk 2]
```diff
@@ -122,12 +125,9 @@ fn build_function_arn(account_id: &str, region: &str, function_name: &str) -> St
...
 fn main() -> Result<()> {
     // First load the configuration
-    let lambda_directory = match env::var("LAMBDA_TASK_ROOT") {
-        Ok(val) => val,
-        Err(_) => "/var/task".to_string(),
-    };
-    let config = match config::get_config(Path::new(&lambda_directory)) {
-        Ok(config) => Arc::new(config),
+    let lambda_directory = env::var("LAMBDA_TASK_ROOT").unwrap_or_else(|_| "/var/task".to_string());
+    let env_config = match config::get_config(Path::new(&lambda_directory)) {
```
Comment (duncanista) [inline bottlecap/src/bin/bottlecap/main.rs:129]: This is not technically `env_config` since it includes also `YAML`, should we find a better name for this?

File: bottlecap/src/secrets/decrypt.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+type ResolveFn = fn(String) -> Result<String, Error>;
+
+pub fn resolve_secrets(config: Config, decoder: ResolveFn) -> Result<Config, String> {
+    if !config.api_key.is_empty() {
+        debug!("DD_API_KEY found, not trying to resolve secrets");
+        Ok(config)
+    } else {
+        if !config.api_key_secret_arn.is_empty() {
+            debug!("DD_API_KEY_SECRET_ARN found, trying to resolve ARN secret");
+            let string = config.api_key_secret_arn.clone();
```
Comment (duncanista) [inline bottlecap/src/secrets/decrypt.rs:19]:
```suggestion
            let api_key_secret_arn = config.api_key_secret_arn.clone();
```

File: bottlecap/src/secrets/decrypt.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            let wait_fot_ret = thread::spawn(move || { decoder(string) });
```
Comment (duncanista) [inline bottlecap/src/secrets/decrypt.rs:20]:
```suggestion
            let wait_for_ret = thread::spawn(move || { decoder(string) });
```
what does `ret` mean?

File: bottlecap/src/secrets/decrypt.rs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            match wait_fot_ret.join() {
+                Ok(Ok(secret)) => {
```
Comment (duncanista) [inline bottlecap/src/secrets/decrypt.rs:22]:
```suggestion
                Ok(Ok(secret)) => {
```
I'm pretty sure this is not a great pattern for this 🤔

File: bottlecap/src/secrets/decrypt.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+                Err(e) => {
+                    if let Some(s) = e.downcast_ref::<&str>() {
+                        Err(format!("Thread panicked with message: {}", s))
+                    } else if let Some(s) = e.downcast_ref::<String>() {
+                        Err(format!("Thread panicked with message: {}", s))
+                    } else {
+                        Err("Thread panicked".to_string())
+                    }
+                }
```
Comment (duncanista) [inline bottlecap/src/secrets/decrypt.rs:38]: Isn't there a better way of handling this?

File: scripts/bottlecap/build.sh
Comment (duncanista) [inline scripts/bottlecap/build.sh:1]: Should this be on a different PR?

File: bottlecap-run/runBottlecap.sh
Comment (alexgallotta) [reply bottlecap-run/runBottlecap.sh:1]: I think that is not really used much, so I didn't want to open another PR for only that

File: bottlecap/Cargo.toml (hunk 1 above)
Comment (alexgallotta) [reply bottlecap/Cargo.toml:28]: Probably not, this PR is still very much WIP now, since 2 approaches are being checked.
I will convert it to draft and clean it up before we re evaluate merging it

File: bottlecap/src/bin/bottlecap/main.rs (hunk 2 above)
Comment (alexgallotta) [reply bottlecap/src/bin/bottlecap/main.rs:135]: Good point, I am thinking that not the config but only the secret will be returned before we merge this. I will rename it once I remove this extra config

File: bottlecap/src/secrets/decrypt.rs (hunk 4 above)
Comment (alexgallotta) [reply bottlecap/src/secrets/decrypt.rs:28]: wait for return, but this code is definitely useless. There was a lot of back and forth and I will refactor before next submission
